### PR TITLE
Implementing the Empty activity in bpel process visualization

### PR DESCRIPTION
**Pull request by NatashaWso2**

### CODE REVIEW

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java [hunk 1]
```diff
@@ -70,4 +83,240 @@ public Element getSVGString(SVGDocument doc) {
...
+            ActivityInterface activity = null;
+            String id = null;
+            ActivityInterface seqActivity = null;
+            SVGCoordinates myStartCoords = getStartIconExitArrowCoords();
+            SVGCoordinates myExitCoords = getEndIconEntryArrowCoords();
+            SVGCoordinates exitCoords = null;
+            SVGCoordinates activityEntryCoords = null;
+            SVGCoordinates activityExitCoords = null;
+            Iterator<ActivityInterface> itr = subActivities.iterator();
+            Element subGroup = doc.createElementNS("http://www.w3.org/2000/svg", "g");
```
Comment (amalhub) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java]: String "http://www.w3.org/2000/svg" has been used in several places. Create constants and reference where necessary.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java [hunk 2]
```diff
@@ -691,4 +716,49 @@ protected Element getImageDefinition(SVGDocument doc, String imgPath, int imgXLe
...
+        return getStartEndImageDef(doc, getStartIconPath(), getStartIconXLeft(),
+                getStartIconYTop(), getStartIconWidth(), getStartIconHeight(),
+                getStartImageId());
+    }
+
+    protected Element getStartEndImageDef(SVGDocument doc, String imgPath, int imgXLeft, int imgYTop,
+                                          int imgWidth, int imgHeight, String id) {
+
+        Element group = null;
+        group = doc.createElementNS("http://www.w3.org/2000/svg", "g");
```
Comment (amalhub) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java]: String "http://www.w3.org/2000/svg" has been used in several places. Create constants and reference where necessary.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/SourceImpl.java [hunk 3]
```diff
@@ -49,4 +91,63 @@ public SourceImpl(OMElement omElement, ActivityInterface parent) {
...
+        return getStartEndImageDef(doc, getStartIconPath(), getStartIconXLeft(),
+                getStartIconYTop(), getStartIconWidth(), getStartIconHeight(),
+                getStartImageId());
+    }
+
+    protected Element getStartEndImageDef(SVGDocument doc, String imgPath, int imgXLeft, int imgYTop,
+                                          int imgWidth, int imgHeight, String id) {
+
+        Element group = null;
+        group = doc.createElementNS("http://www.w3.org/2000/svg", "g");
```
Comment (amalhub) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/SourceImpl.java]: String "http://www.w3.org/2000/svg" has been used in several places. Create constants and reference where necessary.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java (hunk 1 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java]: Fixed with 512bfae

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java (hunk 2 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java]: Fixed with 512bfae

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/SourceImpl.java (hunk 3 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/SourceImpl.java]: Fixed with 512bfae

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/EmptyInterface.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2010, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/EmptyInterface.java]: Incorrect year. If you added this it cannot be 2010 :)

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ActivityImpl.java [hunk 5]
```diff
@@ -658,13 +686,13 @@ protected String getArrowStyle() {
 
     protected String getLinkArrowStyle() {
         String largeArrowStr =
-                "fill:none;fill-rule:evenodd;stroke:#000000;stroke-width:1.5;stroke-linecap:"
+                "fill:none;fill-rule:evenodd;stroke:#FF0000;stroke-width:3;stroke-linecap:"
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ActivityImpl.java]: These should go to a constants class.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseIfImpl.java [hunk 6]
```diff
@@ -125,6 +132,7 @@ public void layoutVertical(int startXLeft, int startYTop) {
 
     }
 
+
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseIfImpl.java]: Format code. There are unnecessary empty lines.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java [hunk 7]
```diff
@@ -70,4 +83,240 @@ public Element getSVGString(SVGDocument doc) {
...
+        if (startX == endX || startY == endY) {
+            path.setAttributeNS(null, "d", "M " + startX + "," + startY + " L " + endX + "," + endY);
+        } else {
+            if (layoutManager.isVerticalLayout()) {
+                path.setAttributeNS(null, "d", "M " + startX + "," + startY + " L " + startX + "," +
+                        ((startY + 2 * endY) / 3) + " L " + endX + "," + ((startY + 2 * endY) / 3) + " L " + endX +
+                        "," + endY);
+            } else {
+                path.setAttributeNS(null, "d", "M " + startX + "," + startY + " L " + ((startX + 1 * endX) / 2) +
+                        "," + startY + " L " + ((startX + 1 * endX) / 2) + "," + endY + " L " + endX + "," + endY);                              //use constants for these propotions
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java]: Use constants for these propotions

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/EmptyImpl.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2010, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/EmptyImpl.java]: Incorrect year. Fix all such issues.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java [hunk 9]
```diff
@@ -428,26 +446,96 @@ protected Element getArrows(SVGDocument doc) {
...
+
                     if (prevActivity != null && prevActivity instanceof ElseIfImpl) {
                         exitCoords = ((ElseIfInterface) prevActivity).getNextElseExitArrowCoords();
                         id = prevActivity.getId() + "-" + activity.getId();
-                        subGroup.appendChild(getArrowDefinition(doc, exitCoords.getXLeft(), exitCoords.getYTop(), activityEntryCoords.getXLeft(), activityEntryCoords.getYTop(), id));
-                        subGroup.appendChild(getArrowDefinition(doc, activityExitCoords.getXLeft(), activityExitCoords.getYTop(), myExitCoords.getXLeft(), myExitCoords.getYTop(), id));
+                        if (activity instanceof ElseImpl) {
+                            boolean check = ((ElseImpl) activity).throwOrNot;
+                            subGroup.appendChild(getArrowDefinition(doc, exitCoords.getXLeft(), exitCoords.getYTop(), activityEntryCoords.getXLeft() - getEndIconWidth() / 2, exitCoords.getYTop(), id));
+                            if (check == true) {
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java]: Nothing done inside if block. use '!=' instead.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java [hunk 10]
```diff
@@ -428,26 +446,96 @@ protected Element getArrows(SVGDocument doc) {
...
+                            } else {
+                                subGroup.appendChild(getArrowDefinition(doc, activityExitCoords.getXLeft(), activityExitCoords.getYTop(), myExitCoords.getXLeft(), myExitCoords.getYTop(), id));
+                            }
+                        } else if (activity instanceof ElseIfImpl) {
+                            boolean check = ((ElseIfImpl) activity).throwOrNot;
+                            subGroup.appendChild(getArrowDefinition(doc, exitCoords.getXLeft(), exitCoords.getYTop(), activityEntryCoords.getXLeft(), activityEntryCoords.getYTop(), id));
+                            if (check == true) {
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java]: if not used

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java [hunk 11]
```diff
@@ -428,26 +446,96 @@ protected Element getArrows(SVGDocument doc) {
...
+                        }
+                    } else if (prevActivity instanceof ThrowImpl && activity instanceof ElseImpl) {
+                        subGroup.appendChild(getArrowDefinition(doc, myStartElseCoords.getXLeft(), myStartElseCoords.getYTop(), activityEntryCoords.getXLeft(), activityEntryCoords.getYTop(), id));
+                    } else if (activity instanceof ElseIfImpl) {
+                        boolean check = ((ElseIfImpl) activity).throwOrNot;
+                        subGroup.appendChild(getArrowDefinition(doc, myStartElseCoords.getXLeft(), myStartElseCoords.getYTop(), activityEntryCoords.getXLeft(), activityEntryCoords.getYTop(), id));
 
-                    } else {
+                        if (check == true) {
+                            //No exit arrow . Process stops from there
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java]: nothing done in if ??

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java [hunk 12]
```diff
@@ -508,7 +530,9 @@ public Element getSVGString(SVGDocument doc) {
         }
         // Get Sub Activities
         group.appendChild(getSubActivitiesSVGString(doc));
+
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java]: Why these spaces?

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java [hunk 13]
```diff
@@ -691,4 +716,49 @@ protected Element getImageDefinition(SVGDocument doc, String imgPath, int imgXLe
...
+            Element x = null;
+            x = doc.createElementNS("http://www.w3.org/2000/svg", "g");
+            x.setAttributeNS(null, "id", id);
+
+            int embedImageX = imgXLeft + imgWidth;
+            int embedImageY = imgYTop + (5 / 2);
+
+            Element embedImage = doc.createElementNS("http://www.w3.org/2000/svg", "line");
+            embedImage.setAttributeNS(null, "x1", String.valueOf(imgXLeft + 10));
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java]: Use constants for these strings. You can have a common class like "bpel2SvgCostants" and include all there.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/SequenceImpl.java [hunk 14]
```diff
@@ -242,7 +246,19 @@ protected Element getArrows(SVGDocument doc) {
                     exitCoords = prevActivity.getExitArrowCoords();
                     entryCoords = activity.getEntryArrowCoords();
                     id = prevActivity.getId() + "-" + activity.getId();
-                    subGroup.appendChild(getArrowDefinition(doc, exitCoords.getXLeft(), exitCoords.getYTop(), entryCoords.getXLeft(), entryCoords.getYTop(), id));
+
+                    if (activity instanceof ThrowImpl) {
+                        setCheck(true);
+                    }
+                    if (prevActivity instanceof ThrowImpl) {
+                        //No exit arrow . Process stops from there
```
Comment (vinok88) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/SequenceImpl.java]: this is not necessary.

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java (hunk 13 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java]: Fixed with commit 512bfaecb2904b99f080649ec2cb617d20266d39

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/EmptyInterface.java (hunk 4 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/EmptyInterface.java]: Fixed with commit 9e33b679aedc58406d44d9ae16279d4ba0ee6e1c

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ActivityImpl.java (hunk 5 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ActivityImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseIfImpl.java (hunk 6 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseIfImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java (hunk 7 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/EmptyImpl.java (hunk 8 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/EmptyImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java (hunk 9 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java (hunk 10 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java (hunk 11 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/IfImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java (hunk 12 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ScopeImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/SequenceImpl.java (hunk 14 above)
Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/SequenceImpl.java]: Fixed with commit 9e33b67

File: components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java [hunk 15]
```diff
@@ -70,4 +83,240 @@ public Element getSVGString(SVGDocument doc) {
...
+        setStartIconXLeft(xLeft);
+        setStartIconYTop(yTop);
+        setStartIconTextXLeft(startXLeft + BOX_MARGIN);
+        setStartIconTextYTop(startYTop + BOX_MARGIN + BPEL2SVGFactory.TEXT_ADJUST);
+        getDimensions().setXLeft(startXLeft);
+        getDimensions().setYTop(startYTop);
+
+    }
+
+    public void layoutHorizontal(int startXLeft, int startYTop) {
```
Comment (pamod) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java]: Any reason for not adding JavaDocs ?

Comment (NatashaWso2) [inline components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java]: Fixed and merged 
https://github.com/wso2/carbon-business-process/blob/master/components/bpel/org.wso2.carbon.bpel.ui/src/main/java/org/wso2/carbon/bpel/ui/bpel2svg/impl/ElseImpl.java#L227

**Pull request merged**